feat(route-diagram): Add route statistics into the diagram

mmelko opened the pull request:
Added route diagram stats that are periodically refreshed:
<img width="1008" alt="Screenshot 2023-04-05 at 23 02 20" src="https://user-images.githubusercontent.com/6814482/230212269-b883c04f-b199-47d3-a336-005f3378d2cc.png">

TBD:
 - add tests for the parsing the xml 
 - set the refresh period according hawtio settings 
 - add values for the edges

tadayosi commented on packages/hawtio/src/plugins/camel/route-diagram/routeDiagram.css:
```diff
@@ -32,7 +32,12 @@
   align-items: center;
 }
-.react-flow__node-camel .camelNodeLabel {
+.nodeTooltip {
```
Using camel case for css names is rare. What about using names with hyphens?  i.e. `node-tooltip`

tadayosi commented on packages/hawtio/src/plugins/camel/route-diagram/RouteDiagram.tsx:
```diff
@@ -16,6 +16,8 @@ import { CamelNodeData, visualizationService } from '@hawtiosrc/plugins/camel/ro
 import 'reactflow/dist/style.css'
 import './routeDiagram.css'
 import { Page } from '@patternfly/react-core'
+import { routesService } from '@hawtiosrc/plugins/camel/routes-service'
```
it's in the same plugin module, so you can use relative path `../routes-service`.

tadayosi commented on packages/hawtio/src/plugins/camel/route-diagram/visualization-service.ts:
```diff
@@ -85,22 +87,47 @@ class VisualizationService {
-    const camelNodes = nodes.map(node => ({ id: node.id, data: node, position: { x: 0, y: 0 }, type: 'camel' }))
+    //parse stats
+    const camelNodes = nodes.map(node => {
+      return {
```
no need for `return`

joshiraez replied:
If there is a TS error when trying to return an object, put it between parentheses

https://ultimatecourses.com/blog/return-object-arrow-function

tadayosi commented on packages/hawtio/src/plugins/camel/routes-service.ts:
```diff
@@ -20,6 +20,55 @@ export type CamelRoute = {
+export type ProcessorStats = {
+  id: string
+  index: number | string
+  state: string | null
```
can `string | null` be simply optional, i.e. `state?: string`?

tadayosi commented on packages/hawtio/src/plugins/camel/routes-service.ts:
```diff
@@ -20,6 +20,55 @@ export type CamelRoute = {
+  resetTimestamp: number | string
+  firstExchangeCompleteTimestamp?: number | null
+  firstExchangeCompletedExchangeId?: number | null
+  lastExchangeCompleteTimestamp?: number | null
```
it doesn't seem to make sense to have it optional and type `number | null` at the same time. can't you simply `lastExchangeCompleteTimestamp?: number`?

same for all other variables.

tadayosi commented on packages/hawtio/src/plugins/camel/routes-service.ts:
```diff
@@ -20,6 +20,55 @@ export type CamelRoute = {
+export type RouteStats = {
+  id: string
+  state: string | null
+  sourceLocation: string | null
```
same here

tadayosi commented on packages/hawtio/src/plugins/camel/routes-service.ts:
```diff
@@ -20,6 +20,55 @@ export type CamelRoute = {
+  lastExchangeCompletedTimestamp?: string | null
+  lastExchangeCompletedExchangeId?: string | null
+  lastExchangeFailureTimestamp?: string | null
+  lastExchangeFailureExchangeId?: string | null
```
same here

tadayosi commented on packages/hawtio/src/plugins/camel/routes-service.ts:
```diff
@@ -205,6 +256,102 @@ class RoutesService {
+  async dumbRoutesStatsXML(routesNode: MBeanNode | null): Promise<string | null> {
+    if (!routesNode) {
+      throw new Error('No route/routes node provided')
+    }
```
if it doesn't accept `null` `routesNode`, you shouldn't make it `MBeanNode | null`. and you can remove the if guard.

tadayosi commented on packages/hawtio/src/plugins/camel/routes-service.ts:
```diff
@@ -205,6 +256,102 @@ class RoutesService {
+      redeliveries: parseInt(pDoc.getAttribute('redeliveries') ?? '0'),
+      resetTimestamp: parseInt(pDoc.getAttribute('resetTimestamp') ?? '0'),
+      sourceLineNumber: parseInt(pDoc.getAttribute('sourceLineNumber') ?? '0'),
+      startTimestamp: parseInt(pDoc.getAttribute('startTimestamp') ?? '0'),
```
oh I didn't know the notation `??`. it seems we should prefer this one over `||` in other places as well in most cases.

joshiraez replied:
Neither did I, and after investigating a bit its really handy to avoid some edge cases with falsy values 😄

tadayosi reviewed (commented):
overall it looks great. there are only a few minor styling comments.

phantomjinx commented on packages/hawtio/src/plugins/camel/routes-service.ts:
```diff
@@ -205,6 +256,102 @@ class RoutesService {
     await jolokiaService.execute(objName, 'remove()')
   }
+
+  async dumbRoutesStatsXML(routesNode: MBeanNode | null): Promise<string | null> {
```
Should this be `dump`??

mmelko replied:
whoops.

phantomjinx commented on packages/hawtio/src/plugins/camel/routes-service.ts:
```diff
@@ -205,6 +256,102 @@ class RoutesService {
+
+    return {
+      id: pDoc.id,
+      deltaProcessingTime: parseInt(pDoc.getAttribute('deltaProcessingTime') ?? '0'),
```
Since these parseInt functions are all the same, can they be replaced with a single shorter function?

eg.
```
parseAttr(pDoc: any, name: string) {
  pDoc.getAttribute(name) ....
}
```

phantomjinx commented on packages/hawtio/src/plugins/camel/routes-service.ts:
```diff
@@ -205,6 +256,102 @@ class RoutesService {
+      startTimestamp: parseInt(pDoc.getAttribute('startTimestamp') ?? '0'),
+      state: pDoc.getAttribute('state'),
+      totalProcessingTime: parseInt(pDoc.getAttribute('totalProcessingTime') ?? '0'),
+      firstExchangeCompletedExchangeId: pDoc.getAttribute('firstExchangeCompletedExchangeId'),
```
Likewise, it seems that maybe a loop over `pDoc` attributes would be less verbose?

joshiraez replied:
A more cleaner way to avoid a lot of boilerplate, taking into account refactoring the parseInt part to its own function.

```
Const obj = Object.fromEntries([attributeName,
attributeName2,
...].map(attribute => getFromPdoc(attribute))

Obj.id =pdoc.id

Return obj
```

If there are any special case can be added to the object like with the id.


Anothwr way

```
Const attributesToRetrieve =[attr1, attr2, attr3...]

Return {
id: p.id
...Object.fromEntries(attributesToRetrieve.map(getFromPdoc))
```

joshiraez commented on packages/hawtio/src/plugins/camel/routes-service.ts:
```diff
@@ -205,6 +256,102 @@ class RoutesService {
+    let xml = null
+    let mbeanName = routesNode.getProperty(contextNodeType)
+
+    try {
```
Put the logic inside named functions instead of using a comment imo

mmelko replied:
@joshiraez Thanks for the suggestion, but I don't think I understand. The comment here is just for explanation why I'm checking mBeanName.

joshiraez replied:
I feel like it's a bit hard to understand. I was thinking about putting them in their own functions to avoid the comment.

I have a small question though, you execute with an empty string for the mbeanName in case there is no object name for route node? Or you just catch the error?

In any case, personally I'd try to make clearer the logic for it as it's really hard to understand the if else. The operations are very similar and the intention is not clear. What do you think about something like this?

```
const routeNodeOperation = dumpRouteStatsAsXml
const routesFolderOperation = dumpRoutesStatsAsXml

const mbeanName = routesNode.getProperty(contextNodeType)

const operationForMBean = mbeanName ? routesFolderOperation : routeNodeOperation
const mbeanToQuery = mbeanName ? mbeanName :  routesNode.objectName

try {
  const xml = await jolokiaService.execute(mbeanToQuery ?? '', operationForMBean, [true, true])
} catch (error) {
  throw new Error('Failed to dump routes stats from mbean: ' + mbeanName + error)
}

if (!xml) {
    throw new Error('Failed to extract any xml from mbean: ' + mbeanName)
}

return xml as string
```

joshiraez replied:
Using the ternary in mbeanToQuery because the ?? operator only works for void types. This way I preserve the original logic, you don't need to add a comment as it can be seen that it depends on if mbean is set or not, avoid the if-else and avoid the use of let. What do you think?

![image](https://user-images.githubusercontent.com/17336236/231678054-44892b89-8707-4c0e-9370-8ea5bc35e547.png)

joshiraez replied:
(If you don't like it we can leave it as is, it's just that I found it a bit confusing at first discerning the logic with the comment)

joshiraez commented:
Tidbits:

* Shouldnt this be a draft?
* Any way to just comment a PR without blocking it? Does any comment create a conversation thus blocking it?

mmelko commented:
@tadayosi @phantomjinx I did some changes here. I changed the types for the attributes of statistics to the strings. That makes it easier to work with. If we will do some math things we can convert later.

I also added the tests covering parsing. I still haven't added Inflight statistic to the connection between the nodes, however it's visible in the small table on hover so I guess it's not crucial here. It can be done later. 

I also did some preparation for showing all the statistics not only the one as in old hawtio but I have to think about it more how to display it properly. However I believe if you are ok with this it can be merged and rest can be tweaked with future PRs.

tadayosi reviewed (approved):
There are still some comments that are not yet addressed, but I think it's ok to merge.

mmelko commented:
UPDATE: I added inflights exchanges number, just not tied to the connection edge but rather to the node.  ![Screenshot 2023-04-13 at 07 56 27](https://user-images.githubusercontent.com/6814482/231666800-15ee243b-cceb-4961-867a-0e734a1c887d.png)
